TOOL-527 Updated user fraction param to be freely configurable within…

richard-bogdan-bitrise opened the pull request:
… 0 and 1.

lpusok commented on step.yml:
```diff
@@ -114,12 +111,7 @@ inputs:
       description: |-
         __Only available if `track` is `rollout`.__
 
-        Portion of the users who should get the staged rollout version of the app.
```
From a user experience point, I feel that the predefined option list is a better option, as it would reduce errors due to misunderstanding the description.
The option to add a different option in the yml can be restored, so if a different number than in the list is needed can be provided by advanced users.
IMO add every tenth fraction from 0.1, 0.2, ... ,1.0

richard-bogdan-bitrise replied:
Added inputs with 5%, please check.

Note: I have to do a dep update after the merge of https://github.com/bitrise-io/go-steputils/pull/14

lpusok commented on step.yml:
```diff
@@ -114,13 +111,28 @@ inputs:
+        - "0.8"
+        - "0.85"
+        - "0.9"
+        - "0.95"
```
It could be an issue if the user wants to release to all users (so no staging); can they do that at the moment? See https://developers.google.com/android-publisher/tracks.

richard-bogdan-bitrise replied:
It is not said explicitly on the docs, but I expect to use release status for that (set it to completed) and do not define user fraction. See APK Workflow example here:
https://developers.google.com/android-publisher/tracks

Also docs says this to fraction:
Fraction of users who are eligible to receive the release. 0 < fraction < 1. To be set, release status must be "inProgress" or "halted".
Please see https://developers.google.com/android-publisher/api-ref/edits/tracks

I think fraction should not be a required input then, I update the code accordingly.

lpusok commented on step.yml:
```diff
@@ -99,40 +99,18 @@ inputs:
-        - "0.85"
-        - "0.9"
-        - "0.95"
+        Portion of the users who should get the staged version of the app. Accepts values between 0.0 and 1.0.
```
Some users would think that they can enter 1 (`between` could be understood like that), add an example how to set it to "100%"

richard-bogdan-bitrise replied:
Updated description.

lpusok commented on main.go:
```diff
@@ -390,7 +389,7 @@ func main() {
 		case betaTrackName:
 			possibleTrackNamesToUpdate = []string{alphaTrackName}
-		case rolloutTrackName, productionTrackName:
+		case productionTrackName:
```
Do not know if this 'Deactivating blocking apk versions' is still necessary or maybe the internal track needs to be added here also.

trapacska commented on config.go:
```diff
@@ -18,7 +18,7 @@ type Configs struct {
 	ExpansionfilePath       string          `env:"expansionfile_path"`
 	Track                   string          `env:"track,required"`
-	UserFraction            string          `env:"user_fraction,opt[0.05,0.1,0.2,0.5]"`
+	UserFraction            string          `env:"user_fraction,range]0.0..1.0["`
```
could you please pass the api description of this range?

richard-bogdan-bitrise replied:
Sure:
https://developers.google.com/android-publisher/api-ref/edits/tracks
See releases[].userFraction

trapacska commented on config.go:
```diff
@@ -18,7 +18,7 @@ type Configs struct {
 	ExpansionfilePath       string          `env:"expansionfile_path"`
 	Track                   string          `env:"track,required"`
-	UserFraction            string          `env:"user_fraction,opt[0.05,0.1,0.2,0.5]"`
+	UserFraction            string          `env:"user_fraction,range]0.0..1.0["`
```
the type is string, AFAIK stepconf will not parse/validate it this way.

richard-bogdan-bitrise replied:
Updated to float

trapacska commented on main.go:
```diff
@@ -340,7 +339,7 @@ func main() {
 
-	if configs.Track == rolloutTrackName {
+	if configs.UserFraction != "" {
 		userFraction, err := strconv.ParseFloat(configs.UserFraction, 64)
```
this will not be required AFAIK

richard-bogdan-bitrise replied:
If it is def value, it should not be assigned.

richard-bogdan-bitrise commented:
Updated to androidpublisher v3, please check the code.

lpusok commented on utility/network.go:
```diff
@@ -0,0 +1,118 @@
+
+		jsonKeySource := jsonKeyPth
+		jsonKeyPth = filepath.Join(tmpDir, "key.json")
+		if err := downloadFileWithRetry(jsonKeySource, jsonKeyPth, 3, 3); err != nil {
```
NOTE: no need need to download to a file, byte[] would work just fine; find reusing jsonKeyPth a bit confusing. Realize this is not new code.

richard-bogdan-bitrise replied:
Updated code.

lpusok commented on utility/network.go:
```diff
@@ -0,0 +1,118 @@
+
+// downloadFileWithRetry calls downloadFile method with a given number of retries and waiting interval between the retries.
+func downloadFileWithRetry(downloadURL, targetPath string, numberOfRetries, waitInterval uint) error {
+	if err := retry.Times(numberOfRetries).Wait(time.Duration(waitInterval) * time.Second).Try(func(attempt uint) error {
```
return retry.Times ... is shorter

richard-bogdan-bitrise replied:
Fixed.

lpusok commented on utility/network.go:
```diff
@@ -0,0 +1,118 @@
+// downloadFileWithRetry calls downloadFile method with a given number of retries and waiting interval between the retries.
+func downloadFileWithRetry(downloadURL, targetPath string, numberOfRetries, waitInterval uint) error {
+	if err := retry.Times(numberOfRetries).Wait(time.Duration(waitInterval) * time.Second).Try(func(attempt uint) error {
+		return downloadFile(downloadURL, targetPath)
```
NOTE: would like to create standardized retry, as for example status code not found does not need retry IMO

richard-bogdan-bitrise replied:
Agree with you, but this should be a general utility method (and should not be placed here), but this time I would say it is out of scope, as initially this was a bugfix.

lpusok replied:
Yes, meant as a general comment, we agreed that the note prefix means that.

lpusok commented on utility/publish.go:
```diff
@@ -0,0 +1,316 @@
+}
+
+// UpdateRelease creates and returns a new release object with the given version codes.
+func UpdateRelease(configs config.Configs, versionCodes googleapi.Int64s, releases *[]*androidpublisher.TrackRelease) error {
```
It would be more clear to avoid passing configs around. Could parse it before passing it to this function; pass in a release and releaesatus

richard-bogdan-bitrise replied:
Multiple values are used from config, methods would be less readable if they would have +3-4 args. I choose to pass config, as in the future we might implement draft and halted release statuses, which would require additional config params.

lpusok commented on utility/publish.go:
```diff
@@ -0,0 +1,316 @@
+// GetRelease gets a release from a track based on it's status.
+func GetRelease(config config.Configs, releases *[]*androidpublisher.TrackRelease) *androidpublisher.TrackRelease {
+	status := GetReleaseStatusFromConfig(config)
+	for _, release := range *releases {
```
PLease add commments to explain this beahaviour of the API (one release per status for all releases)

richard-bogdan-bitrise replied:
Multiple values are used from config, methods would be less readable if they would have +3-4 args. I choose to pass config, as in the future we might implement draft and halted release statuses, which would require additional config params.

richard-bogdan-bitrise replied:
I will do an another round of check for this.

lpusok commented on utility/publish.go:
```diff
@@ -0,0 +1,316 @@
+	log.Infof("New version codes: '%v'", newVersionCodes)
+
+	var cleanTrack bool
+	if len(release.VersionCodes) != len(newVersionCodes) {
```
Why the number of versioncodes matters?

richard-bogdan-bitrise replied:
This way if the number of uploaded apps changed, only the newly uploaded apps will be on the store. This is a legacy code/function.

lpusok commented on utility/publish.go:
```diff
@@ -0,0 +1,316 @@
+			log.Warnf("Shadowing app found, removing current (%v) version, adding new (%v)", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, newVersionCodes[i])
+		} else {
+			log.Warnf("Currently released app with (%v) version is higher than new (%v), app with new version code ignored", currentVersionCodes[i], newVersionCodes[i])
```
Does this need to be a warning?

richard-bogdan-bitrise replied:
No, you are completely right, should be info instead.

richard-bogdan-bitrise commented:
Fixed most of the issues, please check. If ok, I will do a new check for the given comment.

lpusok commented:
ok, so far

hnljp commented:
Do you have a status update on this pull request?

trapacska commented on config/config.go:
```diff
@@ -27,8 +27,8 @@ type Configs struct {
-// validate validates the Configs.
-func (c Configs) validate() error {
+// Validate validates the Configs.
+func (c Configs) Validate() error {
```
why did you convert it to exported method?

trapacska commented on config/config.go:
```diff
@@ -113,24 +114,23 @@ func parseAppList(list string) (apps []string) {
-	}
-
+// AppPaths returns the app to deploy, by prefering .aab files.
+func (c Configs) AppPaths() ([]string, []string) {
```
same question as above

richard-bogdan-bitrise commented:
@hnljp sorry for the late response. PR is still under review, hopefully it will be finished soon.

richard-bogdan-bitrise commented:
Due to request, gathered the major refactor related changes:
Downloadfile: Moved to network.go and refactored to return only the content,
jwtConfigFromJSONKeyFile: moved to network.go
jwtConfigFromP12KeyFile: removed, as it was unused
readLocalisedRecentChanges: moved to publish.go
parseUri: moved to network.go
main:
164-190: moved to method CreateHTTPClient
219-329: moved to method uploadApplications
255-264: moved to method UploadAppAPK in publish.go
268-278: moved to method UploadAppBundle in publish.go
283-309: moved to method UploadExpansionFiles in publish.go
313-325: moved to method UploadExpansionFiles in publish.go
336-489: moved to method updateTracks
363-457: moved to removeBlockingVersionFromRelease and sortAndFilterVersionCodes in publish.go. Minor refactor needed, as some of the untracking logic collided with Google Play rules
460-489: moved to updateListing in publish.go

trapacska commented:
@hnljp the new version is just out now! 🙂 https://github.com/bitrise-steplib/steps-google-play-deploy/releases/tag/2.0.0
